### PR TITLE
Proof of Concept for Image Microservice

### PR DESCRIPTION
**Pull request by humphd**
## Issue This PR Addresses

Part of #1627 

## Type of Change

- [ ] **Bugfix**: Change which fixes an issue
- [x] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Description

To give us something to discuss, and a starting point for further development, I've created a simple image service.

See the README.md for info about using it.

### CODE REVIEW

**Comment by humphd** [comment]
To test this:

```
npm install
npm start
```

Now try going to `http://localhost:4444/?w=1024&t=webp&u=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1593642702909-dec73df255d7%3Fixid%3DMXwxMjA3fDF8MHxwaG90by1wYWdlfHx8fGVufDB8fHw%253D%26ixlib%3Drb-1.2.1%26auto%3Dformat%26fit%3Dcrop%26w%3D1350%26q%3D80` in your browser.

**Comment by humphd** [comment]
First problem: how to deal with all the services under `src/api/*` and how to install things?  I could add another layer of what we already do with `npm install` in the root calling into a `package.json` in `src/api/package.json` and then it can call into `src/api/image/package.json`, etc. but there must be a cleaner way?

**Comment by humphd** [comment]
I notice that [npm 7 is now officially released](https://twitter.com/natfriedman/status/1356738152589758465), and it includes [Workspaces](https://github.com/npm/rfcs/blob/latest/implemented/0026-workspaces.md), which we could use to manage these apis and other package.json files.

The downside is that we force our devs to move to npm 7, which @Metropass has been working on in #1617.  But it might be something we move toward in an upcoming release to get away from all these scripts to manage internal packaging.

**Comment by humphd** [comment]
Unrelated to this work, I came across this standard which is really cool: https://iiif.io/api/image/2.0/

**Comment by raygervais** [comment]
> First problem: how to deal with all the services under `src/api/*` and how to install things? I could add another layer of what we already do with `npm install` in the root calling into a `package.json` in `src/api/package.json` and then it can call into `src/api/image/package.json`, etc. but there must be a cleaner way?

I think the easiest for management is each service get's it own Docker image which can be orchestrated, I'd recommend something simple as this to start: 

```dockerfile
FROM node:alpine-lst AS base

WORKDIR /app

COPY . .

RUN npm install

RUN npm start
```

All this does is simply containerize, but with it now we can start the discussion of allowing docker-compose orchestration duties on a per service level. In my opinion, unless you are directly developing  the service, having a container which contains the node_modules and service itself seems simplest. Thoughts?

File: src/api/image/src/app.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+const cors = require('cors');
+const helmet = require('helmet');
+
+const service = require('./service');
+
+const app = express();
+app.use(helmet());
+app.use(cors());
+
+// TODO: figure out how to do logging for each service...
```
Comment (raygervais) [inline src/api/image/src/app.js:11]: This sounds like a logging service which aggregates ALL service logs to a central location. In this case, I imagine that would be either an HTTP call, or a log rotate which gradually shifts logs from the service to the central location.

Comment (humphd) [reply src/api/image/src/app.js:11]: We have Kibana for this, but aren't actively using it yet.  I'd like all of our services to dump logs into it, so I guess it's just a matter of reading in ENV variables to set the Kibana details (like we do now in the root).

**Comment by humphd** [comment]
RE: your Dockerfile, sounds perfect.  Let's start small and improve it as we go.  Once we have a template, we can use it to build more.

**Comment by humphd** [comment]
@raygervais should I let you do PRs against my branch at this point, or do you want me to do more first?  I'm assuming you'll run with what I have.  Let me know if that's wrong.

**Comment by raygervais** [comment]
I can continue here if that's alright, will add to your commits: 
- Dockerfile
- The basic Traefik configuration under /src/api

If you want to avoid email spam, we could land this item and I work on it from there in a follow up.

**Comment by humphd** [comment]
I don't want to land this because it will break CI atm.  Do PR(s) against my fork for now.

**Comment by raygervais** [comment]
Next Steps: 

- Let's get CI looking green so we can merge?

**Comment by humphd** [comment]
One way to do this is to use workspaces in npm 7, but that would require all our devs to update to latest npm.  I think it might be a good solution overall, but we could easily get this passing CI with some scripts to do the dependent installs, like we do now with other sub-package.jsons.

**Comment by humphd** [comment]
@raygervais FYI, I pushed some updates to this to add the unsplash download logic, etc.  Hopefully that doesn't break you.

**Comment by humphd** [comment]
I added support for smart width/height processing, so it will resize/crop/etc to get the image to cover the available dimensions.  I've also added a simple gallery view to allow seeing all of our backgrounds in development.

File: src/api/image/src/gallery.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+<html lang="en">
+<head>
+  <meta charset="utf-8">
+  <title>Telescope Background Images</title>
+  <meta name="viewport" content="width=device-width, initial-scale=1">
+  <link rel="icon" type="image/png" href="/?t=png&w=200"/>
+  <style>
+    main {
+      display: flex
+      flex-wrap: wrap;
```
Comment (foxviet) [inline src/api/image/src/gallery.js:13]: not sure whether it's gonna work without semi colon?

Comment (humphd) [reply src/api/image/src/gallery.js]: good catch, fixing...

**Comment by humphd** [comment]
This is ready for review.  It's good enough to land and we can iterate on it.

**Comment by humphd** [comment]
Added some Docker updates to get this working properly with `docker-compose`, now able to run the image service + traefik!

Screen Shot 2021-02-08 at 11 57 27 AM

**Comment by HyperTHD** [comment]
Any updates with this?

**Comment by humphd** [comment]
I'm pretty close to having this ready.  I have two efforts happening in parallel:

- Creating a reusable microservice base package, happening in https://github.com/Seneca-CDOT/satellite.  I hope to get an initial version of that published to npm on Friday.
- Creating the Docker, Docker-Compose, ELK (Elastic/Logstach/Kibana) logging, metrics, and monitoring stack.  I did a demo of this yesterday, and it should be usable in some form by the end of this week or early next week.

I'm going to rewrite this code based on Satellite once I publish that package.

**Comment by raygervais** [comment]
@humphd, tag me when you want to discuss next steps and also get me to review :).

**Comment by humphd** [comment]
@raygervais will do.  I'm just fighting with APM route naming in Kibana, then I should be ready.

**Comment by humphd** [comment]
OK, I think this is ready.  I'd like to get this landed so @chrispinkney has a template to use for his work on the User service.

I think we can improve the Docker setup, but this shouldn't block things.

@raygervais, @manekenpix when you have a sec, r?

**Review by manekenpix — DISMISSED** [review]
LGTM :+1: 

I could access:

- `image.docker.localhost/{image, gallery}`
- `elastic.docker.localhost`
- `kibana.docker.localhost` (could see stats in the observability menu)

**Pull request merged**